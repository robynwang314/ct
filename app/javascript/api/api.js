import axios from 'axios'

const countries = {
  index: () => {
    return axios({
      url: 'api/v1/countries',
      method: "GET",
      dataType: "json",
    });
  },
  show: (name) => {
    return axios({
      url: `api/v1/countries/${name}`,
      method: "GET",
      dataType: "json",
    });
  }
}

const api = {
  countries
};

export default api