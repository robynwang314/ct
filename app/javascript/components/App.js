import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { hot } from 'react-hot-loader/root';
import { useCountryContext, CountrySelectionProvider } from './country-context.jsx'
import api from '../api/api.js'
import SearchBar from './SearchBar/search-bar.jsx'
import ChartContainer from './Chart/chart-container.jsx'
import NavBarTabs from "./Nav/nav-bar-tabs.jsx"
import "./app.scss"

export async function getCountries(setCountries) {
  try {
    const response = await api.countries.index()
    if (response) {
      setCountries(response.data)
    }
  } catch (error) {
    console.log(error)
  }
}

export async function getAlertStatus(country, setLoading, setAlertStatus) {
  if (!country) return null;

  try {
    setLoading(true)
    const alert = await api.countries.travel_advisory(country.label)
    if (alert && alert.data) {
      setAlertStatus(alert.data)
    }
  } catch (error) {
    console.log(error)
  }
  finally {
    setLoading(false)
  }
}

export const MainDocument = () => {
  const { string_parameterize, country, countries, setCountries, alertStatus, setAlertStatus, expanded, handleCollapse } = useCountryContext()
  const [loading, setLoading] = useState(false)

  const countryCode = country?.value
  let countryNamesList = []

  useEffect(() => {
    getCountries(setCountries)
    getAlertStatus(country, setLoading, setAlertStatus)
  }, [country])


  if (countries.length > 0) {
    countryNamesList = countries.map((country) => {
      return country.data.iso_short_name
    })

    // separating this out for testing purposes
    if (countryNamesList[17] && countryNamesList[33]) {
      countryNamesList[17] = "United Kingdom"
      countryNamesList[33] = "Moldova, Republic of"
    }

    countryNamesList.unshift("United States")
    countryNamesList.sort()
  }

  return /*!loading && */(
    <>
      <SearchBar countryNamesList={countryNamesList} />
      <br />
      <h2 style={{ marginBottom: '0px', fontWeight: "bold" }}>{country?.label ? country.label : "No country selected"}</h2>
      <h5 style={{ marginBottom: '0', color: "red" }}>Alert Status:</h5> <p style={{ marginTop: '.35%' }}>{alertStatus?.message}</p>
      <ChartContainer />
      <br />
      <NavBarTabs country={country} />
      {country && <button style={{ position: "fixed", zIndex: "1", top: "95%", left: "90%" }} onClick={handleCollapse}>{expanded
        ? "collapse all" : "expand all"}</button>}
      {/* <div style={{ color: "rgb(81, 82, 81)" }} dangerouslySetInnerHTML={{ __html: data?.important_info }} /> */}
    </>
  )
}

const App = () => {
  return (
    <div className="app-container">
      <CountrySelectionProvider>
        <MainDocument />
      </CountrySelectionProvider>
    </div>
  )
}

App.propTypes = {

}

export default hot(App)

