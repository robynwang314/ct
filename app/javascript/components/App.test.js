import React from 'react';
import { render, act, waitFor, fireEvent } from '@testing-library/react'
import userEvent from "@testing-library/user-event";


import App from './App';
import api from '../api/api.js'
import MockedValues from './mockedConsts.js'

jest.mock(
  "./Chart/chart.jsx",
  () =>
    function Chart() {
      return <div>{"Chart"}</div>;
    }
);

jest.mock(
  './Chart/chart-container.jsx',
  () =>
    function StatsContainer() {
      return <div>{"stats container"}</div>;
    }
);

jest.mock(
  "./Nav/nav-bar-tabs.jsx",
  () =>
    function NavBarTabs() {
      return <div>{"Nav Bar Tabs"}</div>
    }
)

jest.mock("axios");

describe("App", () => {
  it('renders empty state', () => {
    const { container } = render(<App />)
    expect(container).toMatchSnapshot()
  });

  describe("MainDocument", () => {
    describe("#getCountries", () => {
      let mockedPromise;

      beforeEach(() => {
        jest.resetAllMocks
      })

      afterEach(() => {
        jest.clearAllMocks();
      })

      it("succeeds", async () => {
        mockedPromise = Promise.resolve({ data: MockedValues.countries });
        jest.spyOn(api.countries, "index").mockReturnValue(mockedPromise);

        render(<App />)

        await act(() => mockedPromise)
        expect(api.countries.index).toBeCalledTimes(1);
      })

      it("console.log errors", async () => {
        mockedPromise = Promise.reject("rejected promise");
        jest.spyOn(api.countries, "index").mockReturnValue(mockedPromise);
        jest.spyOn(console, 'log');

        render(<App />)

        try {
          expect(api.countries.index).toBeCalledTimes(1);
        } catch {
          expect(console.log).toHaveBeenCalled()
        }
      })
    })

    describe("#getAlertStatus", () => {
      let mockedPromise;

      beforeEach(() => {
        mockedPromise = Promise.resolve({ data: MockedValues.countries });
        jest.spyOn(api.countries, "index").mockResolvedValue(mockedPromise);
      })

      afterEach(() => {
        jest.clearAllMocks();
      })

      it("gets the alert status", async () => {
        const newMockedPromise = Promise.resolve({ data: { message: "mock travel_advisory alert" } });
        jest.spyOn(api.countries, 'travel_advisory').mockReturnValue(newMockedPromise)

        const { container, getByText } = render(<App />)

        await act(() => mockedPromise)
        expect(api.countries.index).toBeCalledTimes(1);


        const selectElement = getByText("Select a country");
        userEvent.click(selectElement)

        await waitFor(() => {
          expect(getByText("United States")).toBeInTheDocument()
          expect(container).toMatchSnapshot()
        })

        fireEvent.keyDown(selectElement, { keyCode: 40 })
        fireEvent.keyDown(selectElement, { keyCode: 40 })
        fireEvent.keyDown(selectElement, { keyCode: 13 });

        expect(api.countries.travel_advisory).toHaveBeenCalledWith("Austria");

        await act(() => newMockedPromise)

        expect(getByText("mock travel_advisory alert")).toBeInTheDocument()

        expect(container).toMatchSnapshot()
      })
    })
  })
})