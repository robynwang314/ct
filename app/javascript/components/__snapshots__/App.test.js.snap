// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App MainDocument #getAlertStatus gets the alert status 1`] = `
<div>
  <div
    class="app-container"
  >
    <div
      style="width: 97%; margin: .5%;"
    >
      <div
        class="selectable css-b62m3t-container"
        id="selectable"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-5-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        >
          <span
            id="aria-selection"
          />
          <span
            id="aria-context"
          >
            option Albania focused, 1 of 6. 6 results available. Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
          </span>
        </span>
        <div
          class=" css-1pahdxg-control"
        >
          <div
            class=" css-319lph-ValueContainer"
          >
            <div
              class=" css-14el2xx-placeholder"
              id="react-select-5-placeholder"
            >
              Select a country
            </div>
            <div
              class=" css-6j8wv5-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-controls="react-select-5-listbox"
                aria-describedby="react-select-5-placeholder"
                aria-expanded="true"
                aria-haspopup="true"
                aria-owns="react-select-5-listbox"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-5-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1gtu0rj-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class=" css-26l3qy-menu"
          id="react-select-5-listbox"
        >
          <div
            class=" css-4ljt47-MenuList"
          >
            <div
              aria-disabled="false"
              class=" css-1n7v3ny-option"
              id="react-select-5-option-0"
              tabindex="-1"
            >
              Albania
            </div>
            <div
              aria-disabled="false"
              class=" css-yt9ioa-option"
              id="react-select-5-option-1"
              tabindex="-1"
            >
              Andorra
            </div>
            <div
              aria-disabled="false"
              class=" css-yt9ioa-option"
              id="react-select-5-option-2"
              tabindex="-1"
            >
              Austria
            </div>
            <div
              aria-disabled="false"
              class=" css-yt9ioa-option"
              id="react-select-5-option-3"
              tabindex="-1"
            >
              Germany
            </div>
            <div
              aria-disabled="false"
              class=" css-yt9ioa-option"
              id="react-select-5-option-4"
              tabindex="-1"
            >
              Spain
            </div>
            <div
              aria-disabled="false"
              class=" css-yt9ioa-option"
              id="react-select-5-option-5"
              tabindex="-1"
            >
              United States
            </div>
          </div>
        </div>
      </div>
    </div>
    <br />
    <h2
      style="margin-bottom: 0px; font-weight: bold;"
    >
      No country selected
    </h2>
    <h5
      style="margin-bottom: 0px; color: red;"
    >
      Alert Status:
    </h5>
     
    <p
      style="margin-top: .35%;"
    />
    <div>
      stats container
    </div>
    <br />
    <div>
      Nav Bar Tabs
    </div>
  </div>
</div>
`;

exports[`App MainDocument #getAlertStatus gets the alert status 2`] = `
<div>
  <div
    class="app-container"
  >
    <div
      style="width: 97%; margin: .5%;"
    >
      <div
        class="selectable css-b62m3t-container"
        id="selectable"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-5-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1s2u09g-control"
        >
          <div
            class=" css-319lph-ValueContainer"
          >
            <div
              class=" css-qc6sy-singleValue"
            >
              Austria
            </div>
            <div
              class=" css-6j8wv5-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-5-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 0; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <br />
    <h2
      style="margin-bottom: 0px; font-weight: bold;"
    >
      Austria
    </h2>
    <h5
      style="margin-bottom: 0px; color: red;"
    >
      Alert Status:
    </h5>
     
    <p
      style="margin-top: .35%;"
    >
      mock travel_advisory alert
    </p>
    <div>
      stats container
    </div>
    <br />
    <div>
      Nav Bar Tabs
    </div>
    <button
      style="position: fixed; z-index: 1; top: 95%; left: 90%;"
    >
      collapse all
    </button>
  </div>
</div>
`;

exports[`App renders empty state 1`] = `
<div>
  <div
    class="app-container"
  >
    <div
      style="width: 97%; margin: .5%;"
    >
      <div
        class="selectable css-b62m3t-container"
        id="selectable"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1s2u09g-control"
        >
          <div
            class=" css-319lph-ValueContainer"
          >
            <div
              class=" css-14el2xx-placeholder"
              id="react-select-2-placeholder"
            >
              Select a country
            </div>
            <div
              class=" css-6j8wv5-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <br />
    <h2
      style="margin-bottom: 0px; font-weight: bold;"
    >
      No country selected
    </h2>
    <h5
      style="margin-bottom: 0px; color: red;"
    >
      Alert Status:
    </h5>
     
    <p
      style="margin-top: .35%;"
    />
    <div>
      stats container
    </div>
    <br />
    <div>
      Nav Bar Tabs
    </div>
  </div>
</div>
`;
